### PR TITLE
Fix overflow when reading length, add fuzzer

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
Looks like the project doesn't use rustfmt, is this intentional?

File: fuzz/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+[package.metadata]
+cargo-fuzz = true
+
+[dependencies]
+libfuzzer-sys = "0.4"
+
+[dependencies.yasna]
+path = ".."
+
+# Prevent this from interfering with workspaces
```
Comment (est31) [inline fuzz/Cargo.toml]: Could this be added to the top yasna.rs workspace?

File: fuzz/.gitignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+target
```
Comment (est31) [inline fuzz/.gitignore:2]: target is not needed here in the gitignore if we are using workspaces.

**Review by est31 — CHANGES_REQUESTED** [review]
Thanks for the PR and good catch!

The bugfix and regression test are good to go, my main concerns are about the added fuzzer. First, it seems it's autogenerated, which makes me wonder why it has to be added into git? Once it's in there it has to see at least some level of maintencance, which suddenly is *not* automatable (but running autogeneration at a future point in time *can* be done, but for all of that you don't need it to be in git).

Second, if it is added to git I'd love to have the fuzzer at least *tested* in the CI, that is it doesn't have to run but at least be built. I guess it should only be built on the non-MSRV jobs. Plus I'd love it to be part of the main workspace. Multiple target directories and lockfiles per git project are annoying.

**Comment by est31** [comment]
For rustfmt, it still doesn't have the options stabilized/available that I'd like, so atm it's not mature yet I'd say.

**Comment by ghost** [comment]
The fuzzer is generated from a template, but you can't autogenerate the *whole* thing.

`decode_ber` was hand written as a fairly simple test, and probably *could* be auto generated, but having it actually in git encourages people to actually run it and report issues, even if it's not actually continuously ran anywhere.

Though in practice, yeah, you're not going to be changing the fuzzer's Cargo.toml all that much, except maybe to change what features it uses from the library. 

Personally I don't think it's a great idea to have the fuzzer crate actually as *part* of the main crate, especially since that would be introducing a workspace where there wasn't one before. I've not done that before, and it doesn't really seem like an intended use-case for [`cargo-fuzz`](https://github.com/rust-fuzz/cargo-fuzz).

It would be reasonable to make it built in CI, though.

**Comment by est31** [comment]
> Personally I don't think it's a great idea to have the fuzzer crate actually as part of the main crate

That's not my suggestion. My suggestion is to have a workspace with two crates. Then the Cargo.lock is shared as well as the target dir. Only one directory to clean.

**Comment by est31** [comment]
> The fuzzer is generated from a template, but you can't autogenerate the whole thing.

I see. It can stay then. :+1:

**Review by est31 — APPROVED** [review]
Thanks a lot!

**Pull request merged**